File: langs/java/src/main/java/dev/dotlyte/Validator.java
package dev.dotlyte;

import java.util.*;
import java.util.regex.Pattern;

/**
 * Schema validation engine for DOTLYTE v2.
 */
public final class Validator {

    private static final Map<String, Pattern> FORMAT_PATTERNS = new LinkedHashMap<>();
    static {
        FORMAT_PATTERNS.put("email",
                Pattern.compile("^[a-zA-Z0-9._%+\\-]+@[a-zA-Z0-9.\\-]+\\.[a-zA-Z]{2,}$"));
        FORMAT_PATTERNS.put("uuid",
                Pattern.compile("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"));
        FORMAT_PATTERNS.put("date",
                Pattern.compile("^\\d{4}-\\d{2}-\\d{2}$"));
        FORMAT_PATTERNS.put("ipv4",
                Pattern.compile("^\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}$"));
    }

    private Validator() {}

    /**
     * Validate config data against a schema.
     */
    @SuppressWarnings("unchecked")
    public static List<SchemaViolation> validate(
            Map<String, Object> data,
            Map<String, SchemaRule> schema,
            boolean strict) {

        List<SchemaViolation> violations = new ArrayList<>();

        for (Map.Entry<String, SchemaRule> entry : schema.entrySet()) {
            String key = entry.getKey();
            SchemaRule rule = entry.getValue();
            Object val = getNestedValue(data, key);

            if (val == null) {
                if (rule.isRequired()) {
                    violations.add(new SchemaViolation(key,
                            "required key '" + key + "' is missing", "required"));
                }
                continue;
            }

            // Type check
            if (rule.getType() != null && !checkType(val, rule.getType())) {
                violations.add(new SchemaViolation(key,
                        "expected type '" + rule.getType() + "', got " + val.getClass().getSimpleName(),
                        "type"));
            }

            // Format check
            if (rule.getFormat() != null && val instanceof String) {
                if (!checkFormat((String) val, rule.getFormat())) {
                    violations.add(new SchemaViolation(key,
                            "value '" + val + "' does not match format '" + rule.getFormat() + "'",
                            "format"));
                }
            }

            // Pattern check
            if (rule.getPattern() != null && val instanceof String) {
                if (!Pattern.matches(rule.getPattern(), (String) val)) {
                    violations.add(new SchemaViolation(key,
                            "value '" + val + "' does not match pattern '" + rule.getPattern() + "'",
                            "pattern"));
                }
            }

            // Enum check
            if (rule.getEnumValues() != null && !rule.getEnumValues().contains(val)) {
                violations.add(new SchemaViolation(key,
                        "value " + val + " not in allowed values: " + rule.getEnumValues(),
                        "enum"));
            }

            // Min/Max
            if (val instanceof Number) {
                double num = ((Number) val).doubleValue();
                if (rule.getMin() != null && num < rule.getMin()) {
                    violations.add(new SchemaViolation(key,
                            "value " + num + " is less than minimum " + rule.getMin(),
                            "min"));
                }
                if (rule.getMax() != null && num > rule.getMax()) {
                    violations.add(new SchemaViolation(key,
                            "value " + num + " is greater than maximum " + rule.getMax(),
                            "max"));
                }
            }
        }

        // Strict mode
        if (strict) {
            Set<String> flatKeys = flattenKeys(data, "");
            for (String k : flatKeys) {
                if (!schema.containsKey(k)) {
                    violations.add(new SchemaViolation(k,
                            "unknown key '" + k + "' (strict mode)", "strict"));
                }
            }
        }

        return violations;
    }

    /**
     * Apply schema defaults.
     */
    public static void applyDefaults(
            Map<String, Object> data,
            Map<String, SchemaRule> schema) {
        for (Map.Entry<String, SchemaRule> entry : schema.entrySet()) {
            if (entry.getValue().getDefaultValue() != null) {
                if (getNestedValue(data, entry.getKey()) == null) {
                    setNestedValue(data, entry.getKey(), entry.getValue().getDefaultValue());
                }
            }
        }
    }

    /**
     * Get all sensitive keys from schema.
     */
    public static List<String> getSensitiveKeys(Map<String, SchemaRule> schema) {
        List<String> keys = new ArrayList<>();
        for (Map.Entry<String, SchemaRule> entry : schema.entrySet()) {
            if (entry.getValue().isSensitive()) {
                keys.add(entry.getKey());
            }
        }
        return keys;
    }

    /**
     * Assert valid — throws if violations exist.
     */
    public static void assertValid(
            Map<String, Object> data,
            Map<String, SchemaRule> schema,
            boolean strict) {
        List<SchemaViolation> violations = validate(data, schema, strict);
        if (!violations.isEmpty()) {
            StringBuilder sb = new StringBuilder("Schema validation failed:\n");
            for (SchemaViolation v : violations) {
                sb.append("  - ").append(v).append("\n");
            }
            throw new DotlyteException.ValidationException(violations);
        }
    }

    // ── Helpers ──────────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    static Object getNestedValue(Map<String, Object> data, String key) {
        String[] parts = key.split("\\.");
        Object current = data;
        for (String part : parts) {
            if (current instanceof Map) {
                current = ((Map<String, Object>) current).get(part);
                if (current == null) return null;
            } else {
                return null;
            }
        }
        return current;
    }

    @SuppressWarnings("unchecked")
    static void setNestedValue(Map<String, Object> data, String key, Object value) {
        String[] parts = key.split("\\.");
        Map<String, Object> current = data;
        for (int i = 0; i < parts.length - 1; i++) {
            current.putIfAbsent(parts[i], new LinkedHashMap<>());
            Object next = current.get(parts[i]);
            if (next instanceof Map) {
                current = (Map<String, Object>) next;
            } else {
                Map<String, Object> newMap = new LinkedHashMap<>();
                current.put(parts[i], newMap);
                current = newMap;
            }
        }
        current.put(parts[parts.length - 1], value);
    }

    @SuppressWarnings("unchecked")
    private static Set<String> flattenKeys(Map<String, Object> data, String prefix) {
        Set<String> keys = new LinkedHashSet<>();
        for (Map.Entry<String, Object> entry : data.entrySet()) {
            String fullKey = prefix.isEmpty() ? entry.getKey() : prefix + "." + entry.getKey();
            if (entry.getValue() instanceof Map) {
                keys.addAll(flattenKeys((Map<String, Object>) entry.getValue(), fullKey));
            } else {
                keys.add(fullKey);
            }
        }
        return keys;
    }

    private static boolean checkType(Object val, String expected) {
        switch (expected) {
            case "string": return val instanceof String;
            case "number": return val instanceof Number;
            case "boolean": return val instanceof Boolean;
            case "array": return val instanceof List || val instanceof Object[];
            case "object": return val instanceof Map;
            default: return true;
        }
    }

    private static boolean checkFormat(String val, String format) {
        switch (format) {
            case "url":
                return val.startsWith("http://") || val.startsWith("https://");
            case "ip":
            case "ipv4":
                return FORMAT_PATTERNS.get("ipv4").matcher(val).matches();
            case "ipv6":
                return val.contains(":") && val.length() > 2;
            case "port":
                try {
                    int p = Integer.parseInt(val);
                    return p >= 1 && p <= 65535;
                } catch (NumberFormatException e) { return false; }
            case "email":
                return FORMAT_PATTERNS.get("email").matcher(val).matches();
            case "uuid":
                return FORMAT_PATTERNS.get("uuid").matcher(val).matches();
            case "date":
                return FORMAT_PATTERNS.get("date").matcher(val).matches();
            default:
                return Pattern.matches(format, val);
        }
    }
}

File: langs/php/src/Validator.php
<?php

declare(strict_types=1);

namespace Dotlyte;

/**
 * Schema validation engine for DOTLYTE v2.
 */
final class Validator
{
    private const FORMAT_PATTERNS = [
        'email' => '/^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$/',
        'uuid'  => '/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/',
        'date'  => '/^\d{4}-\d{2}-\d{2}$/',
        'ipv4'  => '/^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/',
    ];

    /**
     * Validate config data against a schema.
     *
     * @param array<string, mixed> $data
     * @param array<string, SchemaRule> $schema
     * @return list<SchemaViolation>
     */
    public static function validate(array $data, array $schema, bool $strict = false): array
    {
        $violations = [];

        foreach ($schema as $key => $rule) {
            $val = self::getNested($data, $key);

            if ($val === null) {
                if ($rule->required) {
                    $violations[] = new SchemaViolation(
                        $key,
                        "required key '{$key}' is missing",
                        'required'
                    );
                }
                continue;
            }

            // Type check
            if ($rule->type !== null && !self::checkType($val, $rule->type)) {
                $type = get_debug_type($val);
                $violations[] = new SchemaViolation(
                    $key,
                    "expected type '{$rule->type}', got {$type}",
                    'type'
                );
            }

            // Format check
            if ($rule->format !== null && is_string($val) && !self::checkFormat($val, $rule->format)) {
                $violations[] = new SchemaViolation(
                    $key,
                    "value '{$val}' does not match format '{$rule->format}'",
                    'format'
                );
            }

            // Pattern check
            if ($rule->pattern !== null && is_string($val)) {
                if (!preg_match($rule->pattern, $val)) {
                    $violations[] = new SchemaViolation(
                        $key,
                        "value '{$val}' does not match pattern '{$rule->pattern}'",
                        'pattern'
                    );
                }
            }

            // Enum check
            if ($rule->enumValues !== null && !in_array($val, $rule->enumValues, true)) {
                $violations[] = new SchemaViolation(
                    $key,
                    "value not in allowed values",
                    'enum'
                );
            }

            // Min/Max
            if (is_int($val) || is_float($val)) {
                if ($rule->min !== null && $val < $rule->min) {
                    $violations[] = new SchemaViolation(
                        $key,
                        "value {$val} is less than minimum {$rule->min}",
                        'min'
                    );
                }
                if ($rule->max !== null && $val > $rule->max) {
                    $violations[] = new SchemaViolation(
                        $key,
                        "value {$val} is greater than maximum {$rule->max}",
                        'max'
                    );
                }
            }
        }

        // Strict mode
        if ($strict) {
            $flatKeys = self::flattenKeys($data);
            foreach ($flatKeys as $k) {
                if (!isset($schema[$k])) {
                    $violations[] = new SchemaViolation(
                        $k,
                        "unknown key '{$k}' (strict mode)",
                        'strict'
                    );
                }
            }
        }

        return $violations;
    }

    /**
     * Apply schema defaults to data.
     *
     * @param array<string, mixed> $data
     * @param array<string, SchemaRule> $schema
     */
    public static function applyDefaults(array &$data, array $schema): void
    {
        foreach ($schema as $key => $rule) {
            if ($rule->defaultValue === null) {
                continue;
            }
            if (self::getNested($data, $key) !== null) {
                continue;
            }
            self::setNested($data, $key, $rule->defaultValue);
        }
    }

    /**
     * Get all sensitive keys from schema.
     *
     * @param array<string, SchemaRule> $schema
     * @return list<string>
     */
    public static function sensitiveKeys(array $schema): array
    {
        $keys = [];
        foreach ($schema as $key => $rule) {
            if ($rule->sensitive) {
                $keys[] = $key;
            }
        }
        return $keys;
    }

    /**
     * Assert valid — throws ValidationException on failure.
     *
     * @param array<string, mixed> $data
     * @param array<string, SchemaRule> $schema
     * @throws ValidationException
     */
    public static function assertValid(array $data, array $schema, bool $strict = false): void
    {
        $violations = self::validate($data, $schema, $strict);
        if (!empty($violations)) {
            $msg = "Schema validation failed:\n" .
                implode("\n", array_map(fn($v) => "  - {$v}", $violations));
            throw new ValidationException($msg, violations: $violations);
        }
    }

    /**
     * Get a nested value via dot-notation key.
     *
     * @param array<string, mixed> $data
     */
    public static function getNested(array $data, string $key): mixed
    {
        $parts = explode('.', $key);
        $current = $data;
        foreach ($parts as $part) {
            if (!is_array($current) || !array_key_exists($part, $current)) {
                return null;
            }
            $current = $current[$part];
        }
        return $current;
    }

    /**
     * Set a nested value via dot-notation key.
     *
     * @param array<string, mixed> $data
     */
    public static function setNested(array &$data, string $key, mixed $value): void
    {
        $parts = explode('.', $key);
        $current = &$data;
        for ($i = 0, $len = count($parts) - 1; $i < $len; $i++) {
            if (!isset($current[$parts[$i]]) || !is_array($current[$parts[$i]])) {
                $current[$parts[$i]] = [];
            }
            $current = &$current[$parts[$i]];
        }
        $current[$parts[count($parts) - 1]] = $value;
    }

    private static function checkType(mixed $val, string $expected): bool
    {
        return match ($expected) {
            'string' => is_string($val),
            'number', 'integer', 'int' => is_int($val) || is_float($val),
            'boolean', 'bool' => is_bool($val),
            'array' => is_array($val),
            default => true,
        };
    }

    private static function checkFormat(string $val, string $fmt): bool
    {
        return match ($fmt) {
            'url' => str_starts_with($val, 'http://') || str_starts_with($val, 'https://'),
            'ip', 'ipv4' => (bool)preg_match(self::FORMAT_PATTERNS['ipv4'], $val),
            'port' => is_numeric($val) && (int)$val >= 1 && (int)$val <= 65535,
            'email' => (bool)preg_match(self::FORMAT_PATTERNS['email'], $val),
            'uuid' => (bool)preg_match(self::FORMAT_PATTERNS['uuid'], $val),
            'date' => (bool)preg_match(self::FORMAT_PATTERNS['date'], $val),
            default => (bool)preg_match("/{$fmt}/", $val),
        };
    }

    /**
     * @param array<string, mixed> $data
     * @return list<string>
     */
    private static function flattenKeys(array $data, string $prefix = ''): array
    {
        $out = [];
        foreach ($data as $key => $value) {
            $fullKey = $prefix === '' ? (string)$key : "{$prefix}.{$key}";
            if (is_array($value)) {
                $out = array_merge($out, self::flattenKeys($value, $fullKey));
            } else {
                $out[] = $fullKey;
            }
        }
        return $out;
    }
}
